packages/abi/src/parser/specifications/v1/resolvable-type.ts

```diff
@@ -0,0 +1,269 @@
+      (tm) => tm.metadataTypeId === metadataTypeId
+    ) as AbiMetadataTypeV1;
+
+    this.metadataType = metadataType;
+    this.type = metadataType.type;
+
+    this.typeParamsArgsMap ??= ResolvableType.mapTypeParametersAndArgs(
+      abi,
+      metadataType,
+      undefined
+    );
+
+    let components = metadataType.components;
+
+    /**
+     * Vectors consist of multiple components,
+     * but we only care about the `buf`'s first type argument
+     * which defines the type of the vector data.
+     * Everything else is being ignored,
+     * as it's then easier to reason about the vector
+     * (you just treat is as a regular struct).
+     */
+    if (swayTypeMatchers.vector(metadataType.type)) {
+      components = components?.map((c) => {
+        if (c.name === 'buf') {
+          return c.typeArguments?.[0];
+        }
+        return c;
+      }) as AbiComponentV1[];
+    }
+
+    this.components = components?.map((c) => ResolvableType.handleComponent(this, abi, c));
+  }
+
+  private static mapTypeParametersAndArgs(
+    abi: AbiSpecificationV1,
+    metadataType: AbiMetadataTypeV1,
+    args: (ResolvableType | ResolvedType)[] | undefined
+  ): Array<[number, ResolvedType | ResolvableType]> | undefined {
+    if (!args) {
+      return metadataType.typeParameters?.map((typeParameter) => [
+        typeParameter,
+        new ResolvableType(abi, typeParameter, undefined),
+      ]);
+    }
+
+    return metadataType.typeParameters?.map((typeParameter, idx) => [typeParameter, args[idx]]);
+  }
+
+  private static handleComponent(
+    parent: ResolvableType,
+    abi: AbiSpecificationV1,
+    c: AbiComponentV1 | AbiTypeArgumentV1
+  ): ResolvableComponent {
+    const name = (c as AbiComponentV1).name;
+
+    if (typeof c.typeId === 'string') {
+      const concreteType = abi.concreteTypes.find(
+        (ct) => ct.concreteTypeId === c.typeId
+      ) as AbiConcreteTypeV1;
+      return {
+        name,
+        type: ResolvableType.resolveConcreteType(abi, concreteType),
+      };
+    }
+
+    const mt = abi.metadataTypes.find((tm) => tm.metadataTypeId === c.typeId) as AbiMetadataTypeV1;
+
+    return {
+      name,
+      type: ResolvableType.handleMetadataType(parent, abi, mt, c.typeArguments),
+    };
+  }
+
+  /**
+   * Concrete types are *resolved* because everything is known about them.
+   */
+  private static resolveConcreteType(
+    abi: AbiSpecificationV1,
+    type: AbiConcreteTypeV1
+  ): ResolvedType {
+    if (type.metadataTypeId === undefined) {
+      return new ResolvedType({
+        type: type.type,
+        typeId: type.concreteTypeId,
+      });
+    }
+
+    if (!type.typeArguments) {
+      return new ResolvableType(abi, type.metadataTypeId, undefined).resolveInternal(
+        type.concreteTypeId,
+        undefined
+      );
+    }
+
+    const metadataType = abi.metadataTypes.find(
+      (mt) => mt.metadataTypeId === type.metadataTypeId
+    ) as AbiMetadataTypeV1;
```

Comment:
I believe we can refactor some code into smaller helper functions. This approach would not only make the code easier to read and understand but also help eliminate repetitive code.

```ts
  private findMetadataType(metadataTypeId: number): AbiMetadataTypeV1 {
    const metadataType = this.abi.metadataTypes.find(
      (type) => type.metadataTypeId === metadataTypeId
    );
    return metadataType;
  }
```

```ts
 private findConcreteType(concreteTypeId: string): AbiConcreteTypeV1 {
    const concreteType = this.abi.concreteTypes.find(
      (type) => type.concreteTypeId === concreteTypeId
    );
    return concreteType;
  }
```
And so on...

Reply:
[2235308](https://github.com/FuelLabs/fuels-ts/pull/3089/commits/2235308a7c210fdcc67c6852e59f0bfd2b905ca6)
[f25c90b](https://github.com/FuelLabs/fuels-ts/pull/3089/commits/f25c90bcc0072a405285f12f0f8ba10ca7feeba9)